### PR TITLE
[WIP] Fixed open declaration completions when there is no identifier

### PR DESCRIPTION
**Pull request by TIHan**
Referring to this issue: https://github.com/Microsoft/visualfsharp/issues/4078

We are getting all symbols when we do completions for open delcarations that do not have an identifier. The reason is because the parser is giving us nothing useful at this specific point.

This fix modifies the parser slightly to give us something to work with. It also gives us a different error message if there is no identifier, "Identifier expected". This is what C# does as well.

#

This needs a bit of discussion. 

In `pars.fsy`, I'm not entirely sure this is a good idea returning: 
```[ SynModuleDecl.Open(Ast.LongIdentWithDots([], []), rhsm) ]```. 

 ```rhsm``` is the range of the OPEN token. Normally the range that is passed to ```SynModuleDecl.Open``` is the entire range of ```Ast.LongIdentWithDots``` when there is an identifier. 

This could be a little confusing. Perhaps there is a better way to represent this kind of thing. I need the range of the OPEN token when there is no identifier in order have completions that result in compilable code.

#

Also, need to add a little more tests.

### CODE REVIEW

File: src/fsharp/pars.fsy [hunk 1]
```diff
@@ -286,7 +286,7 @@ let rangeOfLongIdent(lid:LongIdent) =
 %type <Ast.SynExceptionDefn> exconDefn
 %type <Ast.SynExceptionDefnRepr> exconCore
 %type <Ast.SynModuleDecl list> moduleDefnsOrExprPossiblyEmptyOrBlock
-%type <Ast.LongIdentWithDots> openDecl
+%type <Ast.LongIdentWithDots option> openDecl
```
Comment (vasily-kirichenko) [inline src/fsharp/pars.fsy]: Hmm. Doesn’t this change the F# grammar, allowing empty open statements? If so, is it ok?

**Comment by vasily-kirichenko** [comment]
Having range for open token is nice, I've wanted it on symbols level to gray out the entire "open xxx" in unused open decl analyzer.

File: src/fsharp/pars.fsy (hunk 1 above)
Comment (TIHan) [reply src/fsharp/pars.fsy]: Later on when we check the option, if it is None, we throw an error, "Identifier expected", so it doesn't pass-through.

**Comment by TIHan** [comment]
Maybe ```SynModuleDecl.Open```'s range value should also include the OPEN token range, and/or it should be another value separate: ```Open of tokenRange : range * LongIdentWithDots : longDotId * range : range```

**Comment by vasily-kirichenko** [comment]
Either way would work for my purpose. Would you add this right into this PR?

**Comment by TIHan** [comment]
I've added openTokenRange. The final range is the combined openTokenRange and longDotId.Range.

I'll definitely need some feedback regarding this approach. I feel better adding the openTokenRange, but  AFAIK there isn't anything that indicates we include token ranges anywhere in the syntax tree. So this one is special, wondering if it's appropriate.

**Comment by vasily-kirichenko** [comment]
This is great, thanks!

About tokens in parse tree, they are not there for two reasons, I think: they are not needed _for compilation_, and they should not "pollute" the next compilation stage, as parse tree is not pollute TAST. But, we do need it for tooling, so I’m appreciate this PR.

**Comment by vasily-kirichenko** [comment]
Why it's not merged yet?

**Comment by TIHan** [comment]
Because it's not finished yet. 

Looking at this again, I remember I might need @dsyme's feedback if this is how we should start including keyword ranges into the AST.

**Comment by KevinRansom** [comment]
@TIHan,  is this still an active PR?

**Comment by KevinRansom** [comment]
@TIHan please reopen when you have time to complete this work.

Thanks

Kevin

**Pull request closed without merging**